### PR TITLE
Fix the dependencies for the Windows build

### LINKED ISSUE #47: The Windows build is failing on master (dependencies misaligned)
The windows build is presently set as an `allowed_fail` build but this isn't really the case, as it is generally stable and presently used by Windows users. Since this build uses Anaconda explicitly to get around the nightmare of installing/compiling C-libraries, it's important to note that changes to the requirements (in setup.py or requirements.txt) need to be mirrored in the `environment.yml`, otherwise WIndows-installed pip will try to install dependencies, and we'll get failure stack traces that make no sense at all.

If no one disagrees, I'm going to re-align the dependencies and set the Windows build to be a required passing build.

### PR DESCRIPTION
**Pull request by Zeitsperre**
Fixes #47

### CODE REVIEW

**Review by cehbrecht — APPROVED** [review]

**Comment by cehbrecht** [comment]
@Zeitsperre Thanks for the fix:). roocs-utils is under development ... a package on pypi will follow before a clisops release.

**Comment by Zeitsperre** [comment]
Okay good to know. Thanks!

I'll add a workaround in for now, but once the package is available on conda-forge, it'll make things *much* easier if we can just place it in the `environment.yml` here.

**Comment by Zeitsperre** [comment]
Just waiting on https://github.com/roocs/roocs-utils/pull/14 before I can close this. Thanks for the review @cehbrecht

**Comment by Zeitsperre** [comment]
`OSError: [WinError 145] The directory is not empty`

This appears to be an issue with pip on Windows: https://github.com/pypa/pip/issues/4734

It could be that there are too many additional files loaded by the submodule or the submodule folder names are being corrupted, but in either case, this implies that the only way forward with the current configuration is a conda-forge package of roocs-utils. The doctree is unfortunately too big to be cleaned in an appropriate amount of time (allegedly, windows runs an `rmtree` command several times in the span of a few seconds and fails if there are still folders around).

@cehbrecht Have you ever dealt with this problem after a successful pip install on Windows?

**Comment by cehbrecht** [comment]
> `OSError: [WinError 145] The directory is not empty`
> 
> This appears to be an issue with pip on Windows: [pypa/pip#4734](https://github.com/pypa/pip/issues/4734)
> 
> It could be that there are too many additional files loaded by the submodule or the submodule folder names are being corrupted, but in either case, this implies that the only way forward with the current configuration is a conda-forge package of roocs-utils. The doctree is unfortunately too big to be cleaned in an appropriate amount of time (allegedly, windows runs an `rmtree` command several times in the span of a few seconds and fails if there are still folders around).
> 
> @cehbrecht Have you ever dealt with this problem after a successful pip install on Windows?

@Zeitsperre I have no local Windows for testing.

@Zeitsperre @ellesmith88 From the travis log it has troubles with the `mini-esgf-data` folder. This is a git submodule with data used by the tests. It is both available in clisops and in rook-utils. Can it be installed separately and skipped during the pip install? Not running `git submodule`?

**Comment by Zeitsperre** [comment]
I think the issue is that the git submodule can't be pip-installed (on Windows, anyway; There doesn't seem to be an issue with other OSes). 

I wonder what exactly they are needed for in other packages, assuming that tests are passing for `roocs-utils`. Perhaps if this submodule is only needed to test `roccs-utils`, there may be a way to skip cloning the testing files? 

In the event that the `mini-esgf-data` is needed for standard operations, one potential solution for both would be to configure the `mini-esgf-data` as an [intake](https://intake.readthedocs.io/en/latest/index.html) catalogue. If we can avoid having to use `git` calls in a pip package, we'll be saving the users (and ourselves) from a lot of frustration down the line.

**Comment by Zeitsperre** [comment]
I have a theory on how to address this: It might be as simple as ensuring that the tests and testdata submodules are excluded from the MANIFEST.in. I'm going to give that a try, as tests shouldn't be installed via pip to begin with.

**Comment by Zeitsperre** [comment]
The theory is advancing. I need to read up more on git submodules, but the fact that we aren't installing the tests anymore is getting us further (and likely fixing unrecognized problems that exist in other builds as well).

**Comment by agstephens** [comment]
@Zeitsperre: yesterday, I installed roocs-utils, clisops, daops and rook. And for **each** of them, I installed the `mini-esgf-data` repo so that I could run the tests. I am _sure_ that there is a simpler way. I would propose that we change the way it operates as follows:

1. Decide a common location, maybe a platform-specific 'tempdir', where the _data_ repositories will be put.
2. Provide a python decorator for all test functions/methods that require data from _outside_ the current repo.
3. Apply the decorator to all functions/methods.

The decorator might look like:

```
@with_data_dirs(['mini-esgf-data', 'xclim-testdata'])
def test_something():
     ...
```

This would allow us to use the decorator as a single look-up for checking external data directories exist. This could apply to all cases:

- local data exists on the file system (e.g. '#ceda', '#dkrz')
- data from external repos must be accessible and cached (e.g. 'xclim-testdata')

And we could include something at the command line, or the environment, to force a `git pull` from that repo.

The decorator could be written once in `roocs-utils` so that it is available to all.

Tagging: @cehbrecht @ellesmith88

**Comment by agstephens** [comment]
@Zeitsperre I managed to find time to look at the `xclim` approach to accessing test data. Please confirm that I'm looking in the right place. I was looking at:

https://github.com/Ouranosinc/xclim/blob/master/xclim/testing/__init__.py

I have copied it in another codebase that needs the same approach. We are testing it out and I am pretty sure it will work really nicely for us. I'll report back here soon.

**Comment by agstephens** [comment]
@Zeitsperre, I really like your caching approach. I have implemented it for another project and I am now using `pytest fixtures` to manage access to test data files. It is much better than pulling a whole git repo. Thanks!

Before we introduce it in the `roocs` packages, please can you comment on the `TestFile` and `TestDataset` classes in `xclim.testing`: do you use these classes anywhere? I haven't seen them in use, so I'm not sure if they are required. Thanks

**Comment by Zeitsperre** [comment]
Hi @agstephens thanks for the patience, I don't believe that those classes are useful for you here. I think the idea was to use them to not have a cache of testdata files, but they aren't even implemented yet. I would skip them here.

**Comment by agstephens** [comment]
> Hi @agstephens thanks for the patience, I don't believe that those classes are useful for you here. I think the idea was to use them to not have a cache of testdata files, but they aren't even implemented yet. I would skip them here.

Thanks @Zeitsperre, that's great. We can work from that starting point and migrate away from cloning git repositories all over the place 👍

**Comment by Zeitsperre** [comment]
@agstephens I've secured access to a Windows workstation and will be advancing on this in the next while. I may have found one of the of issues with using git submodules already:
```
$ git submodule update --init
Submodule 'tests/mini-esgf-data' (https://github.com/roocs/mini-esgf-data.git) registered for path 'tests/mini-esgf-data'
Submodule 'tests/xclim-testdata' (https://github.com/roocs/xclim-testdata.git) registered for path 'tests/xclim-testdata'
Cloning into 'C:/Users/tresmi1/git/clisops/tests/mini-esgf-data'...
Cloning into 'C:/Users/tresmi1/git/clisops/tests/xclim-testdata'...
error: unable to create file test_data/group_workspaces/jasmin2/cp4cds1/vol1/data/c3s-cordex/output/EUR-11/IPSL/MOHC-HadGEM2-ES/rcp85/r1i1p1/IPSL-WRF381P/v1/day/psl/v20190212/psl_EUR-11_MOHC-HadGEM2-ES_rcp85_r1i1p1_IPSL-WRF381P_v1_day_
20060101-20101231.nc: Filename too long
error: unable to create file test_data/group_workspaces/jasmin2/cp4cds1/vol1/data/c3s-cordex/output/EUR-11/IPSL/MOHC-HadGEM2-ES/rcp85/r1i1p1/IPSL-WRF381P/v1/day/psl/v20190212/psl_EUR-11_MOHC-HadGEM2-ES_rcp85_r1i1p1_IPSL-WRF381P_v1_day_
20110101-20201231.nc: Filename too long
error: unable to create file test_data/group_workspaces/jasmin2/cp4cds1/vol1/data/c3s-cordex/output/EUR-11/IPSL/MOHC-HadGEM2-ES/rcp85/r1i1p1/IPSL-WRF381P/v1/day/psl/v20190212/psl_EUR-11_MOHC-HadGEM2-ES_rcp85_r1i1p1_IPSL-WRF381P_v1_day_
20210101-20301231.nc: Filename too long
error: unable to create file test_data/group_workspaces/jasmin2/cp4cds1/vol1/data/c3s-cordex/output/EUR-11/IPSL/MOHC-HadGEM2-ES/rcp85/r1i1p1/IPSL-WRF381P/v1/day/psl/v20190212/psl_EUR-11_MOHC-HadGEM2-ES_rcp85_r1i1p1_IPSL-WRF381P_v1_day_
20310101-20401231.nc: Filename too long
error: unable to create file test_data/group_workspaces/jasmin2/cp4cds1/vol1/data/c3s-cordex/output/EUR-11/IPSL/MOHC-HadGEM2-ES/rcp85/r1i1p1/IPSL-WRF381P/v1/day/psl/v20190212/psl_EUR-11_MOHC-HadGEM2-ES_rcp85_r1i1p1_IPSL-WRF381P_v1_day_
20410101-20501231.nc: Filename too long
error: unable to create file test_data/group_workspaces/jasmin2/cp4cds1/vol1/data/c3s-cordex/output/EUR-11/IPSL/MOHC-HadGEM2-ES/rcp85/r1i1p1/IPSL-WRF381P/v1/day/psl/v20190212/psl_EUR-11_MOHC-HadGEM2-ES_rcp85_r1i1p1_IPSL-WRF381P_v1_day_
20510101-20601231.nc: Filename too long
error: unable to create file test_data/group_workspaces/jasmin2/cp4cds1/vol1/data/c3s-cordex/output/EUR-11/IPSL/MOHC-HadGEM2-ES/rcp85/r1i1p1/IPSL-WRF381P/v1/day/psl/v20190212/psl_EUR-11_MOHC-HadGEM2-ES_rcp85_r1i1p1_IPSL-WRF381P_v1_day_
20610101-20701231.nc: Filename too long
error: unable to create file test_data/group_workspaces/jasmin2/cp4cds1/vol1/data/c3s-cordex/output/EUR-11/IPSL/MOHC-HadGEM2-ES/rcp85/r1i1p1/IPSL-WRF381P/v1/day/psl/v20190212/psl_EUR-11_MOHC-HadGEM2-ES_rcp85_r1i1p1_IPSL-WRF381P_v1_day_
20710101-20801231.nc: Filename too long
error: unable to create file test_data/group_workspaces/jasmin2/cp4cds1/vol1/data/c3s-cordex/output/EUR-11/IPSL/MOHC-HadGEM2-ES/rcp85/r1i1p1/IPSL-WRF381P/v1/day/psl/v20190212/psl_EUR-11_MOHC-HadGEM2-ES_rcp85_r1i1p1_IPSL-WRF381P_v1_day_
20810101-20901231.nc: Filename too long
error: unable to create file test_data/group_workspaces/jasmin2/cp4cds1/vol1/data/c3s-cordex/output/EUR-11/IPSL/MOHC-HadGEM2-ES/rcp85/r1i1p1/IPSL-WRF381P/v1/day/psl/v20190212/psl_EUR-11_MOHC-HadGEM2-ES_rcp85_r1i1p1_IPSL-WRF381P_v1_day_
20910101-20991201.nc: Filename too long
Submodule path 'tests/xclim-testdata': checked out 'b3ad1c485fef3c26bb987d8e571b97449bdaf86a'
Unable to checkout '3adba701675ff0318f68fcc55cf716b4e52f65db' in submodule path 'tests/mini-esgf-data'
```

It seems as though the filenames are too long ? I didn't know that that could be a problem. Will provide updates as I find out more.

**Comment by Zeitsperre** [comment]
This is good to go. The sole issue is from the length of the C3S files+paths. I'll open this as a new issue.

File: clisops/ops/subset.py [hunk 1]
```diff
@@ -110,8 +112,17 @@ def subset(
 
     """
     # Convert all inputs to Xarray Datasets
+    if ds is None:
+        raise MissingParameterValue
```
Comment (ellesmith88) [inline clisops/ops/subset.py:116]: This should be raised by ` args = utils.map_params(ds, time, area, level)` on new line 128 if `ds is None`, so I'm wondering if it needs to be implemented twice?

Comment (Zeitsperre) [reply clisops/ops/subset.py:116]: You're right, it doesn't need to be there. I'll see what happens if I take it out.

**Comment by Zeitsperre** [comment]
@ellesmith88 I see that the new release is forthcoming! Feel free to merge this PR at your convenience!

**Comment by ellesmith88** [comment]
@Zeitsperre Thanks - I plan to put it in 🙂 @agstephens @cehbrecht Are you happy with this?

**Review by cehbrecht — APPROVED** [review]

**Comment by agstephens** [comment]
@Zeitsperre @cehbrecht This looks good to me. @ellesmith88 Please go ahead and merge.

**Pull request merged**